File: sdk/typescript/src/ab-test.ts
/**
 * IAOEP TypeScript SDK — A/B Test 集成 (Phase 4)
 *
 * 自动决定 Agent 调用走 baseline / candidate 版本.
 *
 * 用法:
 * ```ts
 * import { abTest, abTestContext } from '@iaoep/sdk';
 *
 * const chat = abTest(
 *   { name: 'customer-service-ab' },
 *   async (query: string) => {
 *     const group = abTestContext.currentGroup();
 *     const model = group === 'baseline' ? 'qwen3-turbo' : 'qwen3-max';
 *     return callLLM(model, query);
 *   }
 * );
 * ```
 */

import { AsyncLocalStorage } from 'node:async_hooks';

// ============================================================
// AsyncLocalStorage: 当前异步上下文的 A/B 测试分组
// ============================================================

interface ABState {
  abTestName: string | null;
  group: string | null;
}

const abStorage = new AsyncLocalStorage<ABState>();

export const abTestContext = {
  /** 设置当前上下文的 A/B 测试名 + 分组 */
  set(abTestName: string, group: string): void {
    abStorage.enterWith({ abTestName, group });
  },

  /** 清除 (自动通过 withABTest 调用) */
  clear(): void {
    // AsyncLocalStorage 自动随作用域结束, 这里仅显式退出
    abStorage.enterWith({ abTestName: null, group: null });
  },

  /** 读当前 group (baseline / candidate / null) */
  currentGroup(): string | null {
    const state = abStorage.getStore();
    return state?.group ?? null;
  },

  /** 读当前 A/B 测试名 */
  currentABTestName(): string | null {
    const state = abStorage.getStore();
    return state?.abTestName ?? null;
  },
};

// ============================================================
// ABTestClient — HTTP 调用 evaluator /assign 端点
// ============================================================

export interface ABTestClientOptions {
  baseUrl?: string;
  cacheTtlSeconds?: number;
  timeoutSeconds?: number;
}

interface CacheEntry {
  group: string;
  expireAt: number;
}

export class ABTestClient {
  private baseUrl: string;
  private cacheTtlMs: number;
  private timeoutMs: number;
  private cache = new Map<string, CacheEntry>();

  constructor(options: ABTestClientOptions = {}) {
    this.baseUrl =
      options.baseUrl
      ?? (typeof process !== 'undefined' ? process.env.IAOEP_EVALUATOR_BASE_URL : undefined)
      ?? 'http://localhost:8082';
    this.cacheTtlMs = (options.cacheTtlSeconds ?? 30) * 1000;
    this.timeoutMs = (options.timeoutSeconds ?? 5) * 1000;
  }

  /**
   * Phase 5: 按 trace_id 细粒度分配 (同一 trace 所有 span 一致).
   */
  async assignByTraceId(abTestName: string, projectId: string, traceId: string): Promise<string> {
    return this.doAssign(abTestName, projectId, `trace:${traceId}`);
  }

  /**
   * Phase 4 兼容 API: 按 userId 分配.
   */
  async assign(abTestName: string, projectId: string, userId: string): Promise<string> {
    return this.doAssign(abTestName, projectId, `user:${userId || 'anonymous'}`);
  }

  private async doAssign(abTestName: string, projectId: string, stickyKey: string): Promise<string> {
    const cacheKey = `${abTestName}:${stickyKey}`;
    const now = Date.now();
    const cached = this.cache.get(cacheKey);
    if (cached && cached.expireAt > now) return cached.group;

    let group = 'baseline';
    try {
      const resp = await fetch(
        `${this.baseUrl}/api/v1/iaoep/projects/${projectId}/ab-tests/${abTestName}/assign`,
        { signal: AbortSignal.timeout(this.timeoutMs) }
      );
      if (resp.ok) {
        const data = await resp.json() as { group?: string };
        group = data.group ?? 'baseline';
      }
    } catch (e) {
      console.debug('[iaoep-sdk] ABTest assign failed, fallback to baseline:', e);
    }

    this.cache.set(cacheKey, { group, expireAt: now + this.cacheTtlMs });
    return group;
  }
}

// ============================================================
// abTest() 高阶函数
// ============================================================

let _client: ABTestClient | null = null;

function getClient(): ABTestClient {
  if (!_client) _client = new ABTestClient();
  return _client;
}

export interface ABTestOptions {
  name: string;
  groups?: [string, string];        // 默认 ['baseline', 'candidate']
  projectId?: string;
  cacheSeconds?: number;
}

/**
 * abTest() — 高阶函数包装, 自动决定 baseline / candidate 版本.
 *
 * 装饰 async 函数, 调用前调 evaluator /assign 决定 group, 写入 AsyncLocalStorage.
 * 被装饰函数内部可通过 `abTestContext.currentGroup()` 读到 group.
 *
 * Phase 5: 按 trace_id 细粒度分配 (同一 trace 所有 span 一致).
 */
export function abTest<TArgs extends unknown[], TReturn>(
  options: ABTestOptions,
  fn: (...args: TArgs) => Promise<TReturn>,
): (...args: TArgs) => Promise<TReturn> {
  const client = getClient();
  const projectId = options.projectId ?? '00000000-0000-0000-0000-000000000001';

  return async (...args: TArgs): Promise<TReturn> => {
    const traceId = extractTraceId(args);
    const group = await client.assignByTraceId(options.name, projectId, traceId);

    console.debug(`[iaoep-sdk] ab_test ${options.name} assigned group=${group} for trace_id=${traceId}`);

    // 进入新的 AsyncLocalStorage 上下文, 让 fn 内部能读 group
    return await abStorage.run({ abTestName: options.name, group }, async () => {
      return fn(...args);
    });
  };
}

function extractTraceId(args: unknown[]): string {
  for (const a of args) {
    if (typeof a === 'string' && a) return a;
    if (a && typeof a === 'object') {
      const obj = a as Record<string, unknown>;
      if (typeof obj.traceId === 'string') return obj.traceId;
      if (typeof obj.trace_id === 'string') return obj.trace_id;
      if (typeof obj.userId === 'string') return obj.userId;
      if (typeof obj.user_id === 'string') return obj.user_id;
    }
  }
  // 兜底: 随机生成 (Phase 5 应从 OTEL current span 取)
  return crypto.randomUUID();
}

File: sdk/python/src/iaoep_sdk/ab_test.py
"""IAOEP Python SDK — A/B Test 集成 (Phase 5: 细粒度).

Phase 4 + 5: 让 SDK 自动决定 Agent 调用走 baseline / candidate 版本,
按 trace_id 细粒度分配 (同一 trace 所有 span 落到同一 group).

用法:
    from iaoep_sdk import ab_test, ab_test_context

    @ab_test(name="customer-service-ab")
    def chat(query: str) -> str:
        group = ab_test_context.current_group()
        return call_llm(query, model="qwen3-turbo" if group == "baseline" else "qwen3-max")
"""

from __future__ import annotations

import functools
import logging
import os
import threading
import time
import uuid
from contextvars import ContextVar
from typing import Any, Callable, Dict, Optional, Tuple
from uuid import UUID

import httpx

logger = logging.getLogger("iaoep-sdk")

from __future__ import annotations

import functools
import logging
import os
import threading
import time
from contextvars import ContextVar
from typing import Any, Callable, Dict, Optional, Tuple
from uuid import UUID

import httpx

logger = logging.getLogger("iaoep-sdk")


# ============================================================
# ContextVar: 当前线程的 A/B 测试分组
# ============================================================

_current_group: ContextVar[Optional[str]] = ContextVar("iaoep_ab_group", default=None)
_current_name: ContextVar[Optional[str]] = ContextVar("iaoep_ab_name", default=None)


class ABTestContext:
    """当前线程的 A/B 测试分组 (baseline / candidate).

    用 ContextVar 实现, 跨异步 协程安全.
    """

    @staticmethod
    def set(ab_test_name: str, group: str) -> None:
        _current_name.set(ab_test_name)
        _current_group.set(group)

    @staticmethod
    def clear() -> None:
        _current_name.set(None)
        _current_group.set(None)

    @staticmethod
    def current_group() -> Optional[str]:
        return _current_group.get()

    @staticmethod
    def current_ab_test_name() -> Optional[str]:
        return _current_name.get()


# ============================================================
# ABTestClient — HTTP 调用 evaluator /assign 端点
# ============================================================

class ABTestClient:
    """IAOEP A/B Test HTTP Client.

    缓存分配结果 (按 ab_test_name + user_id 维度), 避免每次请求都打 evaluator.
    """

    def __init__(
        self,
        base_url: Optional[str] = None,
        cache_ttl_seconds: int = 30,
        timeout_seconds: int = 5,
    ):
        self.base_url = (
            base_url
            or os.environ.get("IAOEP_EVALUATOR_BASE_URL")
            or "http://localhost:8082"
        )
        self.cache_ttl_seconds = cache_ttl_seconds
        self.timeout_seconds = timeout_seconds
        self._cache: Dict[str, Tuple[str, float]] = {}
        self._lock = threading.Lock()
        self._http = httpx.Client(timeout=timeout_seconds)

    def assign(self, ab_test_name: str, project_id: UUID, user_id: str) -> str:
        """调 evaluator 决定 baseline / candidate, 带本地缓存."""
        cache_key = f"{ab_test_name}:{user_id or 'anonymous'}"
        now = time.time()
        with self._lock:
            cached = self._cache.get(cache_key)
            if cached and cached[1] > now:
                return cached[0]

        try:
            resp = self._http.get(
                f"{self.base_url}/api/v1/iaoep/projects/{project_id}/ab-tests/{ab_test_name}/assign"
            )
            resp.raise_for_status()
            data = resp.json()
            group = data.get("group", "baseline")
        except Exception as e:
            logger.debug("ABTest assign failed, fallback to baseline: %s", e)
            group = "baseline"

        with self._lock:
            self._cache[cache_key] = (group, now + self.cache_ttl_seconds)
        return group

    def close(self):
        self._http.close()


# ============================================================
# @ab_test 装饰器
# ============================================================

_client: Optional[ABTestClient] = None


def _get_client() -> ABTestClient:
    """获取 (懒加载) ABTestClient 全局单例."""
    global _client
    if _client is None:
        _client = ABTestClient()
    return _client


def ab_test(
    name: str,
    groups: Tuple[str, str] = ("baseline", "candidate"),
    project_id: Optional[UUID] = None,
    cache_seconds: int = 30,
) -> Callable:
    """@ab_test 装饰器 — 自动决定 baseline / candidate 版本.

    用法:
        @ab_test(name="customer-service-ab")
        def chat(query: str) -> str:
            # 读 ABTestContext.current_group() 决定版本
            group = ab_test_context.current_group()  # 或: from iaoep_sdk import ab_test_context
            model = "qwen3-turbo" if group == "baseline" else "qwen3-max"
            return call_llm(model=model, query=query)
    """
    def decorator(func: Callable) -> Callable:
        client = _get_client()
        pid = project_id or UUID("00000000-0000-0000-0000-000000000001")

        @functools.wraps(func)
        def sync_wrapper(*args, **kwargs) -> Any:
            trace_id = _extract_trace_id(args, kwargs)
            group = client.assign_by_trace_id(name, pid, trace_id)
            logger.debug("ab_test %s assigned group=%s for trace_id=%s", name, group, trace_id)
            ABTestContext.set(name, group)
            try:
                return func(*args, **kwargs)
            finally:
                ABTestContext.clear()

        @functools.wraps(func)
        async def async_wrapper(*args, **kwargs) -> Any:
            trace_id = _extract_trace_id(args, kwargs)
            group = client.assign_by_trace_id(name, pid, trace_id)
            logger.debug("ab_test %s assigned group=%s for trace_id=%s", name, group, trace_id)
            ABTestContext.set(name, group)
            try:
                return await func(*args, **kwargs)
            finally:
                ABTestContext.clear()

        import inspect
        if inspect.iscoroutinefunction(func):
            return async_wrapper
        return sync_wrapper

    return decorator


def _extract_trace_id(args: tuple, kwargs: dict) -> str:
    """从参数提取 trace_id (Phase 5: 优先 trace_id, 兜底 user_id).

    真实生产应从 OTEL current span 取 trace_id; 这里简化从 kwargs 取.
    """
    if "trace_id" in kwargs:
        return str(kwargs["trace_id"])
    if "traceId" in kwargs:
        return str(kwargs["traceId"])
    # 兜底: 随机生成 UUID (同一函数调用内一致)
    return str(uuid.uuid4())


def _extract_user_id(args: tuple, kwargs: dict) -> str:
    """从参数提取 user_id."""
    if "user_id" in kwargs:
        return str(kwargs["user_id"])
    if "userId" in kwargs:
        return str(kwargs["userId"])
    for a in args:
        if isinstance(a, str) and a:
            return a
    return "anonymous"


# 模块级 alias (方便 import)
ab_test_context = ABTestContext
